File: backend/app/schemas/workout.py
from datetime import datetime
from typing import List, Optional
from uuid import UUID
from pydantic import BaseModel, ConfigDict, Field


# ----------------------------------------------------
# Exercise Schemas
# ----------------------------------------------------

class ExerciseBase(BaseModel):
    name: str = Field(..., max_length=100)
    primary_muscle: str = Field(..., max_length=50)
    secondary_muscles: Optional[List[str]] = None
    equipment_required: Optional[List[str]] = None
    difficulty: Optional[str] = Field(None, max_length=20)
    category: Optional[str] = Field(None, max_length=50)
    description: Optional[str] = None
    instructions: Optional[str] = None


class ExerciseCreate(ExerciseBase):
    pass


class ExerciseResponse(ExerciseBase):
    id: UUID
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ----------------------------------------------------
# Workout Plan Schemas
# ----------------------------------------------------

class WorkoutPlanExerciseCreate(BaseModel):
    exercise_id: UUID
    day_of_week: Optional[int] = Field(None, ge=1, le=7)
    sets: Optional[int] = Field(3, ge=1, le=20)
    reps: Optional[str] = Field("8-12", max_length=20)
    rest_seconds: Optional[int] = Field(60, ge=0, le=600)
    notes: Optional[str] = None
    order_index: Optional[int] = 0


class WorkoutPlanExerciseResponse(BaseModel):
    id: UUID
    plan_id: UUID
    exercise_id: UUID
    day_of_week: Optional[int] = None
    sets: Optional[int] = None
    reps: Optional[str] = None
    rest_seconds: Optional[int] = None
    notes: Optional[str] = None
    order_index: Optional[int] = None
    exercise: Optional[ExerciseResponse] = None

    model_config = ConfigDict(from_attributes=True)


class WorkoutPlanCreate(BaseModel):
    name: str = Field("Personalized Workout Plan", max_length=100)
    days_per_week: Optional[int] = Field(4, ge=1, le=7)
    exercises: Optional[List[WorkoutPlanExerciseCreate]] = None


class WorkoutPlanResponse(BaseModel):
    id: UUID
    user_id: UUID
    name: str
    days_per_week: Optional[int] = None
    is_active: bool
    ai_generated: bool
    created_at: datetime
    plan_exercises: List[WorkoutPlanExerciseResponse] = []

    model_config = ConfigDict(from_attributes=True)


# ----------------------------------------------------
# Workout Session Log Schemas
# ----------------------------------------------------

class WorkoutLogExerciseCreate(BaseModel):
    exercise_id: UUID
    set_number: int = Field(..., ge=1, le=20)
    reps_completed: Optional[int] = Field(None, ge=0, le=200)
    weight_kg: Optional[float] = Field(None, ge=0.0, le=1000.0)
    rpe: Optional[int] = Field(None, ge=1, le=10)
    notes: Optional[str] = None


class WorkoutLogExerciseResponse(BaseModel):
    id: UUID
    log_id: UUID
    exercise_id: UUID
    set_number: int
    reps_completed: Optional[int] = None
    weight_kg: Optional[float] = None
    rpe: Optional[int] = None
    notes: Optional[str] = None
    exercise: Optional[ExerciseResponse] = None

    model_config = ConfigDict(from_attributes=True)


class WorkoutLogCreate(BaseModel):
    plan_id: Optional[UUID] = None
    started_at: datetime
    ended_at: Optional[datetime] = None
    notes: Optional[str] = None
    logged_exercises: List[WorkoutLogExerciseCreate]


class WorkoutLogResponse(BaseModel):
    id: UUID
    user_id: UUID
    plan_id: Optional[UUID] = None
    started_at: datetime
    ended_at: Optional[datetime] = None
    notes: Optional[str] = None
    created_at: datetime
    logged_exercises: List[WorkoutLogExerciseResponse] = []

    model_config = ConfigDict(from_attributes=True)

File: src/types/workout.ts
export interface Exercise {
  id: string;
  name: string;
  primary_muscle: string;
  secondary_muscles?: string[] | null;
  equipment_required?: string[] | null;
  difficulty?: string | null;
  category?: string | null;
  description?: string | null;
  instructions?: string | null;
  created_at: string;
}

export interface WorkoutPlanExercise {
  id: string;
  plan_id: string;
  exercise_id: string;
  day_of_week?: number | null;
  sets?: number | null;
  reps?: string | null;
  rest_seconds?: number | null;
  notes?: string | null;
  order_index?: number | null;
  exercise?: Exercise | null;
}

export interface WorkoutPlan {
  id: string;
  user_id: string;
  name: string;
  days_per_week?: number | null;
  is_active: boolean;
  ai_generated: boolean;
  created_at: string;
  plan_exercises: WorkoutPlanExercise[];
}

export interface WorkoutPlanCreate {
  name?: string;
  days_per_week?: number;
  equipment?: string[];
}

export interface WorkoutLogExerciseCreate {
  exercise_id: string;
  set_number: number;
  reps_completed?: number;
  weight_kg?: number;
  rpe?: number;
  notes?: string;
}

export interface WorkoutLogCreate {
  plan_id?: string;
  started_at: string;
  ended_at?: string;
  notes?: string;
  logged_exercises: WorkoutLogExerciseCreate[];
}

export interface WorkoutLogExercise {
  id: string;
  log_id: string;
  exercise_id: string;
  set_number: number;
  reps_completed?: number | null;
  weight_kg?: number | null;
  rpe?: number | null;
  notes?: string | null;
  exercise?: Exercise | null;
}

export interface WorkoutLog {
  id: string;
  user_id: string;
  plan_id?: string | null;
  started_at: string;
  ended_at?: string | null;
  notes?: string | null;
  created_at: string;
  logged_exercises: WorkoutLogExercise[];
}
